rllib/core/torch/torch_sarl_trainer.py

```diff
@@ -0,0 +1,172 @@
+            "compute_grads_and_apply_if_needed_fn": (
+                self.compute_grads_and_apply_if_needed
+            ),
+            "compute_loss_fn": self.compute_loss,
+            "compile_results_fn": self.compile_results,
+            "queues": queues,
+        }
+        train_loop_per_worker = construct_train_func(
+            self._training_func,
+            config,
+            fn_arg_name="train_loop_per_worker",
+            discard_returns=True,
+        )
+        backend_executor = BackendExecutor(
+            backend_config=backend_config,
+            num_workers=scaling_config.num_workers,
+            num_cpus_per_worker=scaling_config.num_cpus_per_worker,
+            num_gpus_per_worker=scaling_config.num_gpus_per_worker,
+            max_retries=0,
+        )
+        checkpoint_manager = CheckpointManager(checkpoint_strategy=None, run_dir=None)
+        # Start the remote actors.
+        backend_executor.start(initialization_hook=None)
+        empty_spec = RayDatasetSpec(dataset_or_dict=None)
+        training_iterator = TrainingIterator(
+            backend_executor=backend_executor,
+            backend_config=backend_config,
+            train_func=train_loop_per_worker,
+            dataset_spec=empty_spec,
+            checkpoint_manager=checkpoint_manager,
+            checkpoint=None,
+            checkpoint_strategy=None,
+        )
+        return queues, training_iterator
+
+    @staticmethod
+    def _training_func(config):
+        rl_module_config = config["rl_module_config"]
+        init_rl_module_fn = config["init_rl_module_fn"]
+        init_optimizer_fn = config["init_optimizer_fn"]
+        compute_loss_fn = config["compute_loss_fn"]
+        compute_grads_and_apply_if_needed_fn = config[
+            "compute_grads_and_apply_if_needed_fn"
+        ]
+        compile_results_fn = config["compile_results_fn"]
+        queue = config["queues"][session.get_local_rank()]
+
+        rl_module = init_rl_module_fn(rl_module_config)
+        device = train.torch.get_device()
+        rl_module.to(device)
+        pg = torch.distributed.new_group(list(range(session.get_world_size())))
+        device_id = [session.get_local_rank()] if str(device).startswith("cuda") else []
+        ddp_rl_module = DDP(rl_module, device_ids=device_id, process_group=pg)
+        unwrapped_module = ddp_rl_module.module
+        optimizer = init_optimizer_fn(ddp_rl_module, {})
+        global_size = session.get_world_size()
+        while 1:
+            whole_batch = queue.get()
+            batch_size = np.ceil(len(whole_batch) / global_size)
+            start = batch_size * session.get_local_rank()
+            end = min(start + batch_size, len(whole_batch))
+            batch = whole_batch[int(start) : int(end)]
+            fwd_out = ddp_rl_module(batch, device=device)
+            loss_dict = compute_loss_fn(batch, fwd_out, device=device)
+            info_dict = compute_grads_and_apply_if_needed_fn(
```

Comment:
+1 on Steven's comment to split these up into compute_grads and apply_grads and then call them both here in sequence.